Add tutorials to documentation (#537)
Qiskit/retworkx: docs/source/tutorial/dags.rst

Directed Acyclic Graphs
=======================

A directed acyclic graph is a directed graph which also doesn't contain
any cycles. A cycle is a non-empty trail in which the first and last nodes in
the trail are the same. For example:

.. jupyter-execute::

   cycle_graph = rx.generators.directed_cycle_graph(5)
   mpl_draw(cycle_graph)

is **not** acyclic. While the earlier path graph is acyclic.

In retworkx you can create a new :class:`.PyDiGraph` object that enforces
that no cycle is added by using the ``check_cycle`` constructor property

.. jupyter-execute::

    dag = rx.PyDiGraph(check_cycle=True)

or after a :class:`~retworkx.PyDiGraph` object is created with the
:attr:`~retworkx.PyDiGraph.check_cycle` attribute:

.. jupyter-execute::

    dag.check_cycle = True

You can also check the status of cycle checking by inspecting the
:attr:`~retworkx.PyDiGraph.check_cycle` attribute:

.. jupyter-execute::

    print(dag.check_cycle)

With ``check_cycle`` set to ``True`` whenever a
:meth:`.PyDiGraph.add_edge` or related functions that add an arbitrary edge to
the graph an error will be raised if a cycle would be introduced. This checking

> does come with a noticeable (and potentially large) runtime overhead so it's

Reviewer: I wonder how the cycle checking works and how large the overhead is (what is the computational complexity).

Author: It's basically just a dfs when adding an edge `(u, v)` if it finds a path from `v` to `u` it errors: https://github.com/Qiskit/retworkx/blob/main/src/digraph.rs#L211-L227

There is some optimization because it maintains the search space in the object so as things are added with `add_edge` repeatedly it doesn't need to explicitly do a full traversal every time.

best to only use it if it's strictly necessary. You can also avoid this
overhead by using the :meth:`.PyDiGraph.add_parent` and
:meth:`.PyDiGraph.add_child` as both add a new node and edge simultaneously and
neither can introduce a cycle.

Applications of DAGs
====================

Topological Sorting
-------------------

> A toplogical sort of a directed acyclic graph defined as :math:`G = (V,E)` is

Reviewer: typo: topological

Author: Fixed in: https://github.com/Qiskit/retworkx/pull/537/commits/f7e4421111f967d25a0b2303bb34e5e110fdd8cc

a linear ordering of all its nodes such that if :math:`G` contains an edge
:math:`(u, v)` then :math:`u` appears before `v`. This only works with dags
because if there was a cycle in the graph :math:`G` then it's not possible
to find such a linear ordering.

> A common application of dags is to use a toplogical sort is to schedule a

Reviewer: typo: topological

Author: Fixed in: https://github.com/Qiskit/retworkx/pull/537/commits/f7e4421111f967d25a0b2303bb34e5e110fdd8cc

sequence of jobs or tasks based on their dependencies. These jobs are
represented by nodes and if there is an edge from :math:`u` to :math:`v` if
job :math:`u` must be completed before job :math:`v`. A topological sort
of a directed acyclic graph will give an order in which to perform these
jobs. For example:

.. jupyter-execute::

    # Create a job dag
    dependency_dag = rx.PyDiGraph(check_cycle=True)
    job_a = dependency_dag.add_node("Job A")
    job_b = dependency_dag.add_child(job_a, "Job B", None)
    job_c = dependency_dag.add_child(job_b, "Job C", None)
    job_d = dependency_dag.add_child(job_a, "Job D", None)
    job_e = dependency_dag.add_parent(job_d, "Job E", None)
    job_f = dependency_dag.add_child(job_e, "Job F", None)
    dependency_dag.add_edge(job_a, job_f, None)
    dependency_dag.add_edge(job_c, job_d, None)

>     mpl_draw(

Reviewer: Tip: Graphviz is really good at drawing DAGs, perhaps an opportunity to showcase `graphviz_draw`?

Reviewer: (Later I noticed that you do that below.)

Author: I updated this to use graphviz in: Fixed in: https://github.com/Qiskit/retworkx/pull/537/commits/f7e4421111f967d25a0b2303bb34e5e110fdd8cc graphviz is a lot more legible here so it makes sense to use it.

***********************
Directed Acyclic Graphs
***********************

This tutorial will explore using retworkx to work with directed acyclic graphs
(also known as a dag).

Directed Graph
==============

A directed graph is a graph that is made up of a set of nodes connected by
directed edges (often called arcs). Edges have a directionality which is
different from undirected graphs where edges have no notion of a direction to
them. For example:

.. jupyter-execute::

    import retworkx as rx
    from retworkx.visualization import mpl_draw

    path_graph = rx.generators.directed_path_graph(5)
    mpl_draw(path_graph)

In this example we created a 5 node directed path graph. This shows the
directionality of the edges in the graph with the arrow head pointing to the
target node. Each edge has a source and target

Directed Acyclic Graphs
=======================

A directed acyclic graph is a directed graph which also doesn't contain
any cycles. A cycle is a non-empty trail in which the first and last nodes in

> the trail are the same. For example:

Author: It's more me being pedantic. A path normally has distinct nodes while a trail is just distinct edges (not necessarily distinct nodes) in a sequence (at least according to my undergrad combinatorics textbook that I dusted off to get formal sounding wording here). Wolfram has a decent definition: https://mathworld.wolfram.com/Trail.html. But if this is confusing I can change it to path.

> the trail are the same. For example:

Reviewer: Not all textbooks use these consistently.

What's important is to pick a terminology and stick to it throughout the documentation. If you do, you might want to include the definitions somewhere in the docs, not only for readers, but also for other people who edit documentation.

> the trail are the same. For example:

Author: I added a footnote with the definition of a trail as part of: https://github.com/Qiskit/retworkx/pull/537/commits/6be2f1eeac38803810216fc9251ea719edfe307a

is **not** acyclic. While the earlier path graph is acyclic.

In retworkx you can create a new :class:`.PyDiGraph` object that enforces
that no cycle is added by using the ``check_cycle`` constructor property

.. jupyter-execute::

    dag = rx.PyDiGraph(check_cycle=True)

or after a :class:`~retworkx.PyDiGraph` object is created with the
:attr:`~retworkx.PyDiGraph.check_cycle` attribute:

.. jupyter-execute::

    dag.check_cycle = True

You can also check the status of cycle checking by inspecting the
:attr:`~retworkx.PyDiGraph.check_cycle` attribute:

.. jupyter-execute::

    print(dag.check_cycle)

With ``check_cycle`` set to ``True`` whenever a
:meth:`.PyDiGraph.add_edge` or related functions that add an arbitrary edge to
the graph an error will be raised if a cycle would be introduced. This checking
does come with a noticeable (and potentially large) runtime overhead so it's
best to only use it if it's strictly necessary. You can also avoid this
overhead by using the :meth:`.PyDiGraph.add_parent` and
:meth:`.PyDiGraph.add_child` as both add a new node and edge simultaneously and
neither can introduce a cycle.

Applications of DAGs
====================

> Topological Sorting

Author: Yeah, that's a better title here I'll change it

> Topological Sorting

Author: Done in: https://github.com/Qiskit/retworkx/pull/537/commits/6e3550845426740b6b95450bbb13693f2512cf8d

-------------------

A topological sort of a directed acyclic graph defined as :math:`G = (V,E)` is
a linear ordering of all its nodes such that if :math:`G` contains an edge

> :math:`(u, v)` then :math:`u` appears before `v`. This only works with dags

Author: Hmm, well we have the `PyDAG` class which is just a backwards compatibility alias for `PyDiGraph`, but I think when we've used it in other places we've done `DAG`. I can update this to be all caps and consistent with the other uses in the docs.

> :math:`(u, v)` then :math:`u` appears before `v`. This only works with dags

Author: Done in: https://github.com/Qiskit/retworkx/pull/537/commits/27d59f73d165ebd9adb30e3ab155894b4c5ceb46

***********************
Directed Acyclic Graphs
***********************

This tutorial will explore using retworkx to work with directed acyclic graphs
(also known as DAGs).

Directed Graph
==============

A directed graph is a graph that is made up of a set of nodes connected by
directed edges (often called arcs). Edges have a directionality which is
different from undirected graphs where edges have no notion of a direction to
them. For example:

.. jupyter-execute::

    import retworkx as rx
    from retworkx.visualization import mpl_draw

    path_graph = rx.generators.directed_path_graph(5)
    mpl_draw(path_graph)

In this example we created a 5 node directed path graph. This shows the
directionality of the edges in the graph with the arrow head pointing to the
target node. Each edge has a source and target.

Directed Acyclic Graphs
=======================

A directed acyclic graph is a directed graph which also doesn't contain

> any cycles. A cycle is a non-empty trail [#]_ in which the first and last nodes in

Author: Yeah, this is probably a bug with https://github.com/Qiskit/qiskit_sphinx_theme this is the normal syntax for footnotes in sphinx (https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#footnotes) so it's probably just not configured well for rendering in the theme. I can do something besides a footnote if you have a suggestion.

> any cycles. A cycle is a non-empty trail [#]_ in which the first and last nodes in

Reviewer 2: I think it's fine as is. Hopefully someone will fix the theme at some point.

.. jupyter-execute::

    print(dag.check_cycle)

With ``check_cycle`` set to ``True`` whenever calling
:meth:`.PyDiGraph.add_edge` or related functions that add an arbitrary edge to
the graph an error will be raised if a cycle would be introduced. This checking
does come with a noticeable (and potentially large) runtime overhead so it's
best to only use it if it's strictly necessary. You can also avoid this
overhead by using :meth:`.PyDiGraph.add_parent` and
:meth:`.PyDiGraph.add_child` as both add a new node and edge simultaneously and
neither can introduce a cycle.

.. [#] A trail is a sequence of nodes and edges on the graph were each edge
   is distinct in the walk.

Applications of DAGs
====================

Task Scheduling
---------------

A topological sort of a directed acyclic graph defined as :math:`G = (V,E)` is
a linear ordering of all its nodes such that if :math:`G` contains an edge
:math:`(u, v)` then :math:`u` appears before `v`. This only works with DAGs
because if there was a cycle in the graph :math:`G` then it's not possible
to find such a linear ordering.

A common application of DAGs is to use a topological sort is to schedule a

> sequence of jobs or tasks based on their dependencies. These jobs are
> represented by nodes and if there is an edge from :math:`u` to :math:`v` if
> job :math:`u` must be completed before job :math:`v`. A topological sort
> of a directed acyclic graph will give an order in which to perform these
> jobs. For example:

Reviewer 2: ```suggestion
sequence of jobs or tasks based on their dependencies. Jobs are
represented by nodes and an edge from node :math:`u` to node :math:`v` implies that
job :math:`u` must be completed before job :math:`v`. A topological sort
of a directed acyclic graph will give an order in which to perform these
jobs. For example:
```

    job_f = dependency_dag.add_child(job_e, "Job F", None)
    dependency_dag.add_edge(job_a, job_f, None)
    dependency_dag.add_edge(job_c, job_d, None)

    graphviz_draw(dependency_dag, node_attr_fn=lambda node: {"label": str(node)})

Above we define a DAG with 6 jobs and dependency relationship between these
jobs. Now if we run the :func:`~retworkx.topological_sort` function on the
graph it will return a linear order to execute the jobs that will respect
the dependency releationship.

.. jupyter-execute::

    topo_sorted = rx.topological_sort(dependency_dag)
    # Print job labels
    print([dependency_dag[job_index] for job_index in topo_sorted])

Qiskit's Compiler
-----------------

Another application using directed acyclic graphs is the compiler in
`Qiskit <https://qiskit.org>`__. Qiskit is an SDK for working with
quantum computing. Qiskit's
`compiler <https://qiskit.org/documentation/apidoc/transpiler.html>`__
internally represents a quantum circuit as a
`directed acyclic graph <https://qiskit.org/documentation/stubs/qiskit.dagcircuit.DAGCircuit.html>`__.
Retworkx was originally started to accelerate the performance of the Qiskit

> compiler's use of directed acyclic graphs for the compiler.

Reviewer 2: ```suggestion
compiler's use of directed acyclic graphs.
```

> To examine how Qiskit's we first need to look at a quantum circuit. A

Reviewer 2: ```suggestion
To examine how Qiskit uses DAGs, we first need to look at a quantum circuit. A
```

quantum circuit is a computation routine consisting of coherent quantum

> operations on quantum data. It is an ordered squence of quantum gates,

Reviewer 2: ```suggestion
operations on quantum data. It is an ordered sequence of quantum operations including gates,
```

and ``q_1`` that is stored in ``c_1``.

We can represent this quantum circuit as a directed acyclic graph like Qiskit
does internally with:

.. jupyter-execute::

    dag = rx.PyDiGraph()
    # Input nodes:
    in_nodes = dag.add_nodes_from(["q_0", "q_1", "c_0", "c_1"])
    # Output nodes
    out_nodes = dag.add_nodes_from(["q_0", "q_1", "c_0", "c_1"])
    # Add H gate
    h_gate = dag.add_child(in_nodes[0], "h", "q_0")
    # Add CX Gate
    cx_gate = dag.add_child(h_gate, "cx", "q_0")
    dag.add_edge(in_nodes[1], cx_gate, "q_1")
    # Add measure Gates
    meas_q0 = dag.add_child(cx_gate, "measure", "q_0")
    meas_q1 = dag.add_child(cx_gate, "measure", "q_1")
    # Measure q0 instruction edges
    dag.add_edge(meas_q0, out_nodes[0], "q_0")
    dag.add_edge(in_nodes[2], meas_q0, "c_0")
    dag.add_edge(meas_q0, out_nodes[2], "c_0")
    # Measure q1 instruction edges
    dag.add_edge(meas_q1, out_nodes[1], "q_1")
    dag.add_edge(in_nodes[3], meas_q1, "c_1")
    dag.add_edge(meas_q1, out_nodes[3], "c_1")

    graphviz_draw(
        dag,
        node_attr_fn=lambda node: {"label": str(node)},
        edge_attr_fn=lambda edge: {"label": str(edge)}
    )

> In this representation of the circuit the flow of data through the bits is

Reviewer 2: ```suggestion
In this representation of the circuit, the flow of data through the bits is
```

modeled by edges. The first set of nodes are input nodes and the last set
are output nodes representing the beginning state and end state of each
bit (both classical and quantum). The compiler then runs analysis and

> transformations on this DAG view of a quantum circuit to optimize the

Reviewer 2: ```suggestion
transformations on the DAG representation of a quantum circuit to optimize the
```

quantum circuit so it can be executed on real hardware. For example, a
simple transformation pass is to translate the quantum gates in the circuit

> to the set of gates allowed on a device. In the above example if we were to
> attempt to run that on a QPU that didn't natively support the ``H`` quantum
> gate we'd have to translate that to an equivalent series of instructions
> that the hardware actually supported. A simplified view of how this is
> performed is:

Reviewer 2: ```suggestion
to the set of gates allowed on a device. If we were to attempt to run
the above circuit on a QPU that didn't natively support the ``H`` quantum
gate, we'd first have to translate it to an equivalent series of instructions
that the hardware actually supported. A simplified view of how this is
performed is:
```

                        return 0

                dag.substitute_node_with_subgraph(
                    gate_index,
                    equivalent_subcircuit,
                    map_fn
                )

    graphviz_draw(
        dag,
        node_attr_fn=lambda node: {"label": str(node)},
        edge_attr_fn=lambda edge: {"label": str(edge)}
    )

Another example of how the compiler in Qiskit operates on a DAG is to perform
analysis to find all the instances of single qubit gates that are executed in
series. This series of quantum gates can be analyzed and often simplified
into a shorter sequence of gates. A simplified example of this analysis is:

.. jupyter-execute::

    bit_nodes = {"q_0", "q_1", "c_0", "c_1"}

    def filter_fn(node):
        # Don't collect input or output nodes
        if node in bit_nodes:
            return False
        # Don't include 2 qubit gates
        if node == "cx":
            return False
        # Ignore non-unitary operations
        if node == "measure":
            return False
        return True

    runs = rx.collect_runs(dag, filter_fn)
    print(runs)

With this we have the DAG nodes that make up a series of 1 qubit gates that

> we can analyze and attempt to simplify. Skipping the details of the

Reviewer 2: I like the name drop for `contract_nodes` (since I wrote it :D), but I wonder if this just gets too complicated for the purpose of this tutorial. Perhaps it makes sense to stop at finding the series of 1-qubit gates to simplify (i.e. stop the tutorial here without showing how to actually do the simplification).

Author: I'll defer to you here, I thought it was good to show the full path of finding runs in the circuit and then showing the simplification because I thought it might be confusing to say we can find these things in the DAG easily but not show what we do with it afterwards. But I don't feel strongly about it and if you think this is too much detail we can drop it.

Reviewer 2: Makes sense. I suggested a change just now that I think would make this a little easier to read, while keeping the content.

is **not** acyclic. While the earlier path graph is acyclic.

In retworkx you can create a new :class:`.PyDiGraph` object that enforces
that no cycle is added by using the ``check_cycle`` constructor property

.. jupyter-execute::

    dag = rx.PyDiGraph(check_cycle=True)

or after a :class:`~retworkx.PyDiGraph` object is created with the
:attr:`~retworkx.PyDiGraph.check_cycle` attribute:

.. jupyter-execute::

    dag.check_cycle = True

You can also check the status of cycle checking by inspecting the
:attr:`~retworkx.PyDiGraph.check_cycle` attribute:

.. jupyter-execute::

    print(dag.check_cycle)

With ``check_cycle`` set to ``True``, methods that add edges
to the graph (e.g. :meth:`.PyDiGraph.add_edge`) will raise an error
if a cycle would be introduced. This checking
does come with a noticeable (and potentially large) runtime overhead so it's
best to only use it if it's strictly necessary. You can also avoid this
overhead by using :meth:`.PyDiGraph.add_parent` and
:meth:`.PyDiGraph.add_child` as both add a new node and edge simultaneously and
neither can introduce a cycle.

.. [#] A trail is a sequence of nodes and edges on the graph where each edge

>    is distinct in the walk.

Reviewer 3: Can you put the footnote content inside a `.. note::` as well? It looks awkward and hard to differentiate from the text in the current format. I know it is a limitation of the documentation theme but I think asking Sphinx to explicitly use note would make it look better

Author: Oops I missed this comment. Given that I'd like to get the release out today and CI is almost done I think we can fix this in a follow-up easily enough post release. So I'll leave it for then.

                        return 0

                dag.substitute_node_with_subgraph(
                    gate_index,
                    equivalent_subcircuit,
                    map_fn
                )

    graphviz_draw(
        dag,
        node_attr_fn=lambda node: {"label": str(node)},
        edge_attr_fn=lambda edge: {"label": str(edge)}
    )

Another example of how the compiler in Qiskit operates on a DAG is to perform
analysis to find all the instances of single qubit gates that are executed in
series. This series of quantum gates can be analyzed and often simplified
into a shorter sequence of gates. A simplified example of this analysis is:

.. jupyter-execute::

    bit_nodes = {"q_0", "q_1", "c_0", "c_1"}

    def filter_fn(node):
        # Don't collect input or output nodes
        if node in bit_nodes:
            return False
        # Don't include 2 qubit gates
        if node == "cx":
            return False
        # Ignore non-unitary operations
        if node == "measure":
            return False
        return True

    runs = rx.collect_runs(dag, filter_fn)
    print(runs)

With this we have the DAG nodes that make up a series of 1 qubit gates that

> we can analyze and attempt to simplify. Skipping the details of the
> internals of how the simplification works we wanted to contract these nodes
> to a single gate we could do that with
> :meth:`~retworkx.PyDiGraph.contract_nodes`. For example, if the 3 node sequence
> returned by :func:`~retworkx.collect_runs`, ``['rz(pi/2)', 'sx', 'rz(pi/2)']``,
> were to be simplified to a single gate ``"U"`` it could be done like:

Reviewer 2: ```suggestion
we can analyze and attempt to simplify.

To perform the simplification, we'll use :meth:`~retworkx.PyDiGraph.contract_nodes` to replace a series of nodes (or gates in our case) with a single equivalent node (gate). For example, if the 3 node sequence
returned by :func:`~retworkx.collect_runs`, ``['rz(pi/2)', 'sx', 'rz(pi/2)']``,
were to be simplified to a single gate ``"U"``, it could be done like:
```
